loopy/transform/callable.py

```diff
@@ -130,43 +129,16 @@ def __init__(self, rule_mapping_context, caller_knl,
 
     def map_subscript(self, expr, expn_state):
         if expr.aggregate.name in self.callee_knl.arg_dict:
-            from loopy.symbolic import get_start_subscript_from_sar
-            from loopy.symbolic import simplify_via_aff
             from pymbolic.primitives import Subscript, Variable
+            from pymbolic import substitute
 
             sar = self.callee_arg_to_call_param[expr.aggregate.name]  # SubArrayRef
 
-            callee_arg = self.callee_knl.arg_dict[expr.aggregate.name]
-            if sar.subscript.aggregate.name in self.caller_knl.arg_dict:
-                caller_arg = self.caller_knl.arg_dict[sar.subscript.aggregate.name]
-            else:
-                caller_arg = self.caller_knl.temporary_variables[
-                        sar.subscript.aggregate.name]
-
-            flatten_index = 0
-            for i, idx in enumerate(get_start_subscript_from_sar(sar,
-                    self.caller_knl).index_tuple):
-                flatten_index += idx*caller_arg.dim_tags[i].stride
-
-            flatten_index += sum(
-                idx * self.rec(tag.stride, expn_state)
-                for idx, tag in zip(self.rec(expr.index_tuple, expn_state),
-                                    callee_arg.dim_tags))
-
-            flatten_index = simplify_via_aff(flatten_index)
-
-            new_indices = []
-            for dim_tag in caller_arg.dim_tags:
-                if dim_tag.stride != 0:
-                    ind = flatten_index // dim_tag.stride
-                else:
-                    # argument has 0-stride i.e. doesn't matter how we index into it.
-                    ind = 0
-                flatten_index -= (dim_tag.stride * ind)
-                new_indices.append(ind)
-
-            new_indices = tuple(simplify_using_aff(
-                self.callee_knl, i) for i in new_indices)
+            index_tuple = self.rec(expr.index_tuple, expn_state)
+            subs_map = {iname: idx for idx, iname in
+                    zip(index_tuple, sar.swept_inames)}
```

Comment:
- This is too simple. A `SubArrayRef` could be `[i]: x[i,i]`, and I don't think this would work. But I think substitution is still a much better approach than the flattening that was there before.
- Also, it seems that we don't have test coverage of this for the inliner... we should!

Reply:
It does work. when there's an access like `y[j]` which is the array `x[i, i]`  passed from the caller, then we replace all occurrences of `i` in `x[i, i]` with `j`.

I added a test just in case. Have a look.